### PR TITLE
Panel Dropdown: Added ServerConVar and reworked whole structure

### PR DESCRIPTION
**Pull request by ZenBre4ker**
Okay, Im gonna test those now, but you can already give feedback to the whole restructuring 😄 

Changelog below:
### Added

- Reworked our simplified Dropdowns MakePanel `PANEL:MakeComboBox(data)` version
  - Added possibility to manipulate serverside ConVars with Dropdowns. Just add .serverConvar with the conVarName to the given data similar to .convar
  - .serverConVar and .conVar are also supported
  - data.choices can now be a table containing `{title, value, select, icon, additionalData}`
  - data.selectValue is added, use it instead of data.selectName to choose the value you set
  - data.selectTitle is added and shall replace data.selectName

### Breaking Changes

- Reworked Dropdowns Panel `DComboBoxTTT2` itself
  - `PANEL:AddChoice(title, value, select, icon, data)` now uses the second argument as value string for setting convars, use the fifth argument for special data instead
  - `PANEL:ChooseOption(value, index, ignoreConVar)` no longer chooses the displayed text, but the underlying value, which also sets conVars
- Reworked our simplified Dropdowns MakePanel `PANEL:MakeComboBox(data)` version
  - `data.OnChange(value, additionalData)` is now only called with the two important arguments, the value that e.g. convars are set and the additionalData

### CODE REVIEW

**Review by TimGoll — APPROVED** [review]
LGTM. However, if you allow .conVar and .serverConVar for drop down menus, you should do the same for all other menu items as well

**Comment by ZenBre4ker** [comment]
Yeah true, but wanted to make changes solely on dropdowns. We can allow them on the others later too.

DONT MERGE! As those are only theoretical changes and Im gonna test all menus before merging.

**Comment by ZenBre4ker** [comment]
LGTM, good to go, when tk or histalek agree

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:169]: I don't really like the design on those two parameters here. 
Either use another function specifically for choosing by value which looks up the index and then calls this function with the index or solve it in another way that does not include using two parameters where only one is required, this feels weird.

Just saw that this is the base function for others that already exist.
Please still rewrite them to use eg. the index function as a base and let the others only handle converting the value to the right index.

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:181]: why is this using 3 functions to extract the data from the choice instead of just getting the choice and access the data directly?

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (saibotk) [inline gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:351]: "TTT2F1MenuConVarChangeCallback" is not unique enough, eg. when having the same variable multiple times for whatever reason, only one callback is registered. Maybe add a unique number or identifier.

**Review by saibotk — CHANGES_REQUESTED** [review]

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:169]: I mainly tried to stick to the original, can remove all old behaviour nonetheless:
https://wiki.facepunch.com/gmod/DComboBox:ChooseOption

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:181]: Also old structure
https://wiki.facepunch.com/gmod/DComboBox:OnSelect

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (TimGoll) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:351]: Is this the GMod provided function? You can leave the identifier away if you don't plan on removing the callback again

Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:351]: I need the callback identifier as I remove the function, when the dropdown is inactive.

File: gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua
Comment (ZenBre4ker) [reply gamemodes/terrortown/gamemode/client/cl_vskin/vgui/dcombobox_ttt2.lua:181]: ah 3 functions because like this i can change the structure anytime. those are the getters

**Comment by ZenBre4ker** [comment]
Your comments made me remember, that those functions are deprecated.

But the callback doesnt need another identifier. Thats a possibility, thats just weird. I couldnt really track those unless I do something globally.

**Comment by TimGoll** [comment]
> But the callback doesnt need another identifier. Thats a possibility, thats just weird. I couldnt really track those unless I do something globally.

I agree with you there. This is such a rare edge cases. And even *if* it happens, the worst that happens is that the convar isn't updated

**Comment by ZenBre4ker** [comment]
Actually the convar gets updated, but all but one of the dropdowns  of the same convar isnt refreshed onchange

**Comment by TimGoll** [comment]
> Actually the convar gets updated, but all but one of the dropdowns of the same convar isnt refreshed onchange

This is what I meant. All but one convars in the UI aren't updated.

**Comment by ZenBre4ker** [comment]
Forgot that local variables in a file are also "global" to all dropdowns. So here you go 😄

**Comment by TimGoll** [comment]
@saibotk You said we should rerequest your review here to it is on your todo list. But we can't since you already are requested.

**Review by saibotk — APPROVED** [review]

**Pull request merged**
